### PR TITLE
e2e_test: integrate MistLoadTest to verify streaming

### PR DESCRIPTION
**Pull request by emranemran**
Use MistLoadTest to spawn multiple clients for a specific protocol (e.g. HLS) as viewers at catalyst-two node. The test passes if every client is able to connect and stream successfully. The json files generated by MistLoadTest are parse to verify if streams passed or failed.

### CODE REVIEW

File: test/e2e/e2e_test.go
Comment (iameli) [inline test/e2e/e2e_test.go:266]: Will defer to @leszko here on best practices for tests like this, but should this maybe be returning an error instead of panicking? Seems like a good practice in general... on the other hand I can see how this error would be like "test harness failure" instead of "failing test" and maybe a panic makes sense there.

File: test/e2e/e2e_test.go
Comment (iameli) [inline test/e2e/e2e_test.go:272]: In any event, no sense in doing anything else after a `Fatalf`, which will crash the process immediately

File: test/e2e/e2e_test.go
Comment (iameli) [inline test/e2e/e2e_test.go:233]: Nice! I'm thinking that each protocol should probably be its own test, rather than just having one big loop in here? Minimally, having a separate function for running a protocol test will let you `return err` in the error cases for those instead of having to do awkward `continue`s or `break`s

**Review by iameli — CHANGES_REQUESTED** [review]
Really really nice! Happy it wasn't too terrible to get this working properly - just suggested a few structuring changes

File: test/e2e/e2e_test.go
Comment (emranemran) [reply test/e2e/e2e_test.go:266]: Yeah, I'm inclined to remove the panic and just make it a general "fail". We'd be reviewing logs anyways to see where the failures are - doesn't seem like a panic would be anymore useful here.

File: test/e2e/e2e_test.go
Comment (emranemran) [reply test/e2e/e2e_test.go:272]: Ah I realize that now after looking at the `Fatalf` man page which calls os.Exit right after. Will fix.

File: test/e2e/e2e_test.go
Comment (emranemran) [reply test/e2e/e2e_test.go:233]: Right, I think I got a little lost in trying to do things in the "idiomatic" go style. I'll take a look at refactoring.

Comment (leszko) [reply test/e2e/e2e_test.go:233]: Yeah, it should be a separate test. Also, the very idiomatic in Go is to use a tests [] array. Please check "Introducing table driven tests" in https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

**Review by leszko — CHANGES_REQUESTED** [review]
Added some comments. But two general thoughts:
1. We should create more tests rather than expanding the assertion of the same test. Having more tests give us better information what failed. One single test is hard to analyze in case of failures.
2. I'd try to not cleanup files manually, but rather depend on the built-in 1`t.TempDir()`

File: test/e2e/e2e_test.go
Comment (emranemran) [reply test/e2e/e2e_test.go:256]: `MistLoadTest` will timeout and exit within 30s (default). I'll turn it into a parameter in the next push and verify it does work as expected.

File: test/e2e/e2e_test.go
Comment (emranemran) [reply test/e2e/e2e_test.go:239]: Yeah, I'll have to update `MistLoadUtil` to accept a different path. But agree managing files is not the best way - didn't know about `t.TempDir()` so i'll try to get that to work.

Comment (emranemran) [reply test/e2e/e2e_test.go:239]: PR here to allow output directory: 
https://github.com/DDVTECH/mistserver/pull/58/commits/1341ffbcf865814f7ebbe2bf67632177987f7146

This should let me refactor into something much simpler.

**Comment by emranemran** [comment]
**Updates:** 
* Refactored to use table driven test as suggested by @leszko. We can now run a load test against each protocol (e.g. HLS) independently or all protocols at the same time as defined in the test matrix
* Using t.TempDir() to get simplify code that tracked/deleted results files
* Blocked on merging until @leszko's PR for redirection merges as I need to rebase on that branch to pick up some changes to e2e_test.go

File: test/e2e/e2e_test.go
Comment (leszko) [inline test/e2e/e2e_test.go:309]: I see we have only one test in `tests`, do we plan to add more? If not, then do we need this loop at all? 🤔

File: test/e2e/e2e_test.go
Comment (leszko) [inline test/e2e/e2e_test.go:298]: nit: Could you move this function up? It should be under `results` type definition. In general, it's better to "used" functions should be under "using" functions. To follow the [Stepdown rule](https://github.com/Geeksltd/Programming.Tips/blob/master/docs/methods/stepdown-rule.md).

File: test/e2e/e2e_test.go
Comment (leszko) [inline test/e2e/e2e_test.go:255]:
```suggestion
	defer cmdMistLoadTest.Process.Kill()
	out, err := cmdMistLoadTest.Output()
	require.NoError(t, err)
	if err != nil {
		glog.Fatalf("Failed to start MistLoadTest: %s", err)
	}
	glog.Infof("MistLoadTest stdout: %s", out)
```

**Review by leszko — APPROVED** [review]
Added some minor comments. Other than that, it looks good. Thanks for all the changes @emranemran 👍

File: test/e2e/e2e_test.go
Comment (emranemran) [reply test/e2e/e2e_test.go:309]: Yep, we'll be adding more streaming protocols to test in the future so decided to just have this framework from the get-go instead of refactoring again in the future.

File: test/e2e/e2e_test.go
Comment (emranemran) [reply test/e2e/e2e_test.go]: Fixed.

File: test/e2e/e2e_test.go
Comment (emranemran) [reply test/e2e/e2e_test.go]: Done

**Pull request merged**
